stl/inc/type_traits

```diff
@@ -1483,140 +1483,197 @@ class reference_wrapper;
+struct _Invoker1<_Callable, _Ty1, _Removed_cvref, false, true>
+    : conditional_t<
+          is_base_of_v<typename _Is_member_object_pointer<_Removed_cvref>::_Class_type, remove_reference_t<_Ty1>>,
+          _Invoker_pmd_object,
+          conditional_t<_Is_specialization_v<_Remove_cvref_t<_Ty1>, reference_wrapper>, _Invoker_pmd_refwrap,
+              _Invoker_pmd_pointer>> {}; // pointer to member data
+
+template <class _Callable, class _Ty1, class _Removed_cvref>
+struct _Invoker1<_Callable, _Ty1, _Removed_cvref, false, false> : _Invoker_functor {};
+
+template <class _Callable>
+_CONSTEXPR20 auto invoke(_Callable&& _Obj) noexcept(noexcept(static_cast<_Callable&&>(_Obj)()))
+    -> decltype(static_cast<_Callable&&>(_Obj)()) {
+    return static_cast<_Callable&&>(_Obj)();
+}
+
+template <class _Callable, class _Ty1, class... _Types2>
+_CONSTEXPR20 auto invoke(_Callable&& _Obj, _Ty1&& _Arg1, _Types2&&... _Args2) noexcept(
+    noexcept(_Invoker1<_Callable, _Ty1>::_Call(
+        static_cast<_Callable&&>(_Obj), static_cast<_Ty1&&>(_Arg1), static_cast<_Types2&&>(_Args2)...)))
+    -> decltype(_Invoker1<_Callable, _Ty1>::_Call(
+        static_cast<_Callable&&>(_Obj), static_cast<_Ty1&&>(_Arg1), static_cast<_Types2&&>(_Args2)...)) {
+#if _HAS_IF_CONSTEXPR
+    if constexpr (_Invoker1<_Callable, _Ty1>::_Strategy == _Invoker_strategy::_Functor) {
+        return static_cast<_Callable&&>(_Obj)(static_cast<_Ty1&&>(_Arg1), static_cast<_Types2&&>(_Args2)...);
+    } else if constexpr (_Invoker1<_Callable, _Ty1>::_Strategy == _Invoker_strategy::_Pmf_object) {
+        auto _Pmf = _Obj;
+        return (static_cast<_Ty1&&>(_Arg1).*_Pmf)(static_cast<_Types2&&>(_Args2)...);
+    } else if constexpr (_Invoker1<_Callable, _Ty1>::_Strategy == _Invoker_strategy::_Pmf_refwrap) {
+        auto _Pmf = _Obj;
+        return (static_cast<_Ty1&&>(_Arg1).get().*_Pmf)(static_cast<_Types2&&>(_Args2)...);
+    } else if constexpr (_Invoker1<_Callable, _Ty1>::_Strategy == _Invoker_strategy::_Pmf_pointer) {
+        auto _Pmf = _Obj;
+        return ((*static_cast<_Ty1&&>(_Arg1)).*_Pmf)(static_cast<_Types2&&>(_Args2)...);
+    } else if constexpr (_Invoker1<_Callable, _Ty1>::_Strategy == _Invoker_strategy::_Pmd_object) {
+        auto _Pmd = _Obj;
+        return static_cast<_Ty1&&>(_Arg1).*_Pmd;
+    } else if constexpr (_Invoker1<_Callable, _Ty1>::_Strategy == _Invoker_strategy::_Pmd_refwrap) {
+        auto _Pmd = _Obj;
+        return static_cast<_Ty1&&>(_Arg1).get().*_Pmd;
+    } else {
+        static_assert(_Invoker1<_Callable, _Ty1>::_Strategy == _Invoker_strategy::_Pmd_pointer, "bug in invoke");
+        auto _Pmd = _Obj;
+        return (*static_cast<_Ty1&&>(_Arg1)).*_Pmd;
+    }
+#else // ^^^ _HAS_IF_CONSTEXPR // !_HAS_IF_CONSTEXPR vvv
+    return _Invoker1<_Callable, _Ty1>::_Call(
+        static_cast<_Callable&&>(_Obj), static_cast<_Ty1&&>(_Arg1), static_cast<_Types2&&>(_Args2)...);
+#endif // _HAS_IF_CONSTEXPR
+}
 
 #if _HAS_CXX17

```

Comment:
Should we consider `#define _C_invoke invoke` in C++20 mode to potentially reduce total instantiations when a program calls both `invoke(args...)` and `_C_invoke(args...)`? Or:
```c++
#if _HAS_CXX20
#define _C_INVOKE invoke
#else
#define _C_INVOKE _C_invoke
// define _C_invoke here
#endif
```
and replace `_C_invoke` with `_C_INVOKE` at the callsites?

Reply:
I'm inclined to think that we should implement #51 unconditionally and just have `invoke` be `constexpr`. WG21-P0859 "Core Issue 1581: When are constexpr member functions defined?" was voted in as a Defect Report in Albuquerque, which means that it applies retroactively to C++14 (so we can report any issues as compiler bugs). As for the library side, our general policy is to guard features for Standard modes, unless the feature is non-disruptive and guarding it would increase complexity significantly. So far, our experience with `constexpr` has been that while it can trigger compiler bugs, it almost always does so when the user is attempting to use it in a constant expression, not ordinary calls (that's why we were able to constexprize so much of the STL without the world exploding). Additionally, making a function `constexpr` is non-disruptive to people who aren't expecting it - it introduces no new identifiers, doesn't change lookup or throughput, etc. (The Library's prohibition on `constexpr` strengthening does not make sense anymore.) At most, people might take a dependency on the feature and then notice that other implementations do guard it - but that seems like a minor risk (entirely at compile time).

The throughput cost of having two `invoke` implementations is nonzero, and the complexity cost (horrible macro) is significant, so I think it should just be unconditionally constexpr.

Reply:
Ditto "I shouldn't be trying to trick you into implementing #51, nevermind for now."

Reply:
Reactivating: I'm going to do just the std::invoke itself only part of #51 unconditionally.

Reply:
s/unconditionally/in C++17/g because our rationale is "apply" and friends added in 17